### PR TITLE
Avoid C-style arrays in Kokkos_ScratchSpace.hpp

### PR DESCRIPTION
**Pull request by masterleinad**
Alternative to #3956 and #3957 working around issues with Intel 18.

@ndellingwood Can you please confirm that this also fixes the issue for you?

### CODE REVIEW

**Comment by ndellingwood** [comment]
@masterleinad this passed tests with intel for me

**Comment by ndellingwood** [comment]
@masterleinad I hadn't tested the unit test with cuda before the PR and missed the change needed to prevent cuda build errors, moving the test from ctor to a member function called by the ctor worked for me to resolve the errors:

```diff
diff --git a/core/unit_test/TestTeam.hpp b/core/unit_test/TestTeam.hpp
index aa0af28..a0192ca 100644
--- a/core/unit_test/TestTeam.hpp
+++ b/core/unit_test/TestTeam.hpp
@@ -1538,6 +1538,10 @@ struct TestTeamPolicyHandleByValue {
   using mem_space  = typename ExecSpace::memory_space;
 
   TestTeamPolicyHandleByValue() {
+    test();
+  }
+
+  void test() {
 #if defined(KOKKOS_ENABLE_CXX11_DISPATCH_LAMBDA)
     const int M = 1, N = 1;
     Kokkos::View<scalar **, mem_space> a("a", M, N);
```

**Comment by ndellingwood** [comment]
@masterleinad I pushed the diff above to your `avoid_c_style_scratch_space_intel branch`

**Comment by masterleinad** [comment]
Thanks for fixing the test!

**Review by ndellingwood — APPROVED** [review]
lgtm, these changes pass with intel/18 and avoid code duplication in #3957

File: core/unit_test/TestTeam.hpp [hunk 1]
```diff
@@ -1529,6 +1529,38 @@ struct TestScratchAlignment {
...
+
+  void test() {
+#if defined(KOKKOS_ENABLE_CXX11_DISPATCH_LAMBDA)
+    const int M = 1, N = 1;
+    Kokkos::View<scalar **, mem_space> a("a", M, N);
+    Kokkos::View<scalar **, mem_space> b("b", M, N);
+    Kokkos::deep_copy(a, 0.0);
+    Kokkos::deep_copy(b, 1.0);
+    Kokkos::parallel_for(
+        "test_tphandle_by_value",
```
Comment (dalg24) [inline core/unit_test/TestTeam.hpp:1550]: Typo?  I would not bother with the label anyway.

File: core/unit_test/TestTeam.hpp [hunk 2]
```diff
@@ -1529,6 +1529,38 @@ struct TestScratchAlignment {
 
 }  // namespace
 
+namespace {
+
+template <class ExecSpace>
+struct TestTeamPolicyHandleByValue {
```
Comment (dalg24) [inline core/unit_test/TestTeam.hpp:1535]: How is that test related to the issue you are fixing?

**Review by dalg24 — CHANGES_REQUESTED** [review]

File: core/unit_test/TestTeam.hpp (hunk 2 above)
Comment (masterleinad) [reply core/unit_test/TestTeam.hpp:1535]: It's a reproducer for the issue @ndellingwood was seeing when testing with `Trilinos`.

File: core/unit_test/TestTeam.hpp (hunk 1 above)
Comment (masterleinad) [reply core/unit_test/TestTeam.hpp:1550]: @ndellingwood Can you take care of this?

File: core/unit_test/TestTeam.hpp (hunk 2 above)
Comment (dalg24) [reply core/unit_test/TestTeam.hpp:1535]: How is it related to whether the team handle is passed by value or not?!  :/

Comment (dalg24) [reply core/unit_test/TestTeam.hpp:1535]: https://github.com/kokkos/kokkos/pull/3957#issue-620735358

Comment (masterleinad) [reply core/unit_test/TestTeam.hpp:1535]: I was able to reproduce the problem with `RelWithDebInfo` but didn't see it for `Debug` with `ICC18`. I would suspect that some optimizations lead to wrong assembly...

**Review by dalg24 — APPROVED** [review]
If you remove/fix the label, I'd prefer you avoid closing and reopening the anonymous namespace.

File: core/unit_test/TestTeam.hpp (hunk 2 above)
Comment (ndellingwood) [reply core/unit_test/TestTeam.hpp:1535]: It tests taking the team policy handle by value rather than by reference, in this particular test it seg faults in builds with intel without the changes in Kokkos_ScratchSpace.hpp (I saw with a Serial build, debug mode, and intel/18, did not test more broadly). 
Testing taking the team policy handle by value rather than reference may be done in other unit tests, so there may be some peculiarities about this particular case that are tickling compiler issues, for example if I replaced the lambda body of `a(i, j) += b(i, j);` with `a(i, j) += 1;` the seg fault was not present

File: core/unit_test/TestTeam.hpp (hunk 1 above)
Comment (ndellingwood) [reply core/unit_test/TestTeam.hpp:1550]: @dalg24 @masterleinad I force pushed the update to remove the label and open/close of anonymous namespace; rebased to a single commit

Comment (ndellingwood) [reply core/unit_test/TestTeam.hpp:1550]: hm, my local rebased changes said they successfully pushed to this branch but not showing up here in github

Comment (ndellingwood) [reply core/unit_test/TestTeam.hpp:1550]: @masterleinad I looked at this branch on your fork and the SHA there matches the changes I made locally and force pushed, but not showing up in the PR, any suggestions to help get this PR to update? I'm not sure what I messed up...
https://github.com/masterleinad/kokkos/commit/13d020d773674a0e50b12c954b82604c53b90a46
